### PR TITLE
Adding the AdamW optimizer

### PR DESCRIPTION
**Pull request by atreyamaj**
# Pull Request Template

## Description

Referencing #2478 

Added the AdamW optimizer: Adam but with modified weight decay.

[Source paper](https://arxiv.org/abs/1711.05101)  

All tests in `deepchem/models/tests/test_optimizers.py` are passing.

Starting this PR as a draft PR (as discussed with @rbharath ) because it requires an additional package to be used: [Tensorflow Addons](https://github.com/tensorflow/addons)  
Documentation can be found here: https://www.tensorflow.org/addons

There are additional commits in this PR because it is built on top of my branches containing code for PRs #2479 and #2480

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

@alat-rights @peastman

### CODE REVIEW

**Comment by atreyamaj** [comment]
Test report for `test_optimizers.py`:

Screenshot from 2021-04-17 20-02-00

**Comment by peastman** [comment]
This PR includes some other changes that aren't related to AdamW.  Did you base it on the wrong branch?

**Comment by atreyamaj** [comment]
> This PR includes some other changes that aren't related to AdamW. Did you base it on the wrong branch?

I've built upon the previous PRs so that we can integrate in the following order: 
[Huber loss](https://github.com/deepchem/deepchem/pull/2479) --> [Poisson Loss](https://github.com/deepchem/deepchem/pull/2480) --> AdamW

There are additional commits here which 'inherit' from the Huber loss and poison loss PRs. @peastman

**Comment by peastman** [comment]
In the future, it's best if you create a separate branch for each PR, all based on the main branch.  That keeps them independent of each other so each one can be reviewed, modified, and merged on its own without affecting the others.

**Comment by atreyamaj** [comment]
> In the future, it's best if you create a separate branch for each PR, all based on the main branch. That keeps them independent of each other so each one can be reviewed, modified, and merged on its own without affecting the others.

Thank you, I will keep this in mind. Would you like to me close the Poisson Loss and AdamW PRs and open fresh new ones based off `main`?

**Comment by peastman** [comment]
We can probably deal with it in this case, since they're all small PRs.  But if any changes need to be made to one of the others, you may find it easiest to separate them rather than having to propagate changes to multiple branches.

**Comment by atreyamaj** [comment]
> We can probably deal with it in this case, since they're all small PRs. But if any changes need to be made to one of the others, you may find it easiest to separate them rather than having to propagate changes to multiple branches.

Thank you! I will do so if there are changes to be made.

**Comment by peastman** [comment]
We also should update env.common.yml so TensorFlow Addons will get installed when running CI.

**Comment by atreyamaj** [comment]
> We also should update env.common.yml so TensorFlow Addons will get installed when running CI.

Done! @peastman

**Review by rbharath — APPROVED** [review]
Looks good on my end! Once @peastman is done reviewing we can merge this in

**Comment by rbharath** [comment]
I've just merged in your earlier PR @atreyamaj. Do you want to rebase this PR on top of master to pull in your other changes?

**Comment by atreyamaj** [comment]
> I've just merged in your earlier PR @atreyamaj. Do you want to rebase this PR on top of master to pull in your other changes?

I have rebased it and force-pushed my `adamw` branch to drop the old commits @rbharath 

Would this be okay?

**Review by rbharath — APPROVED** [review]
Ok I think this is good to merge in! Will leave open for a day or two in case anyone else has comments

**Comment by rbharath** [comment]
@atreyamaj Hmm, this looks to have conflicts now. I think I merged in one of your other PRs and as a result the conflicts showed up. Could you take a look and resolve the conflicts?

**Comment by atreyamaj** [comment]
> @atreyamaj Hmm, this looks to have conflicts now. I think I merged in one of your other PRs and as a result the conflicts showed up. Could you take a look and resolve the conflicts?

@rbharath I've resolved the merge conflicts now. Please take a look!

**Review by rbharath — APPROVED** [review]
LGTM! I think the only errors are the dgllife ones
